substrate/frame/treasury/src/migration.rs

```diff
@@ -0,0 +1,107 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Treasury pallet migrations.
+
+use super::*;
+use alloc::collections::BTreeSet;
+use core::marker::PhantomData;
+use frame_support::{defensive, traits::OnRuntimeUpgrade};
+use pallet_balances::WeightInfo;
+
+/// The log target for this pallet.
+const LOG_TARGET: &str = "runtime::treasury";
+
+pub struct CleanupProposals<T, I>(PhantomData<(T, I)>);
+
+impl<T: Config<I> + pallet_balances::Config, I: 'static> OnRuntimeUpgrade
+	for CleanupProposals<T, I>
+{
+	fn on_runtime_upgrade() -> frame_support::weights::Weight {
+		let mut approval_index = BTreeSet::new();
+		for approval in Approvals::<T, I>::get().iter() {
+			approval_index.insert(*approval);
+		}
+
+		let mut proposals_released = 0;
+		for (proposal_index, p) in Proposals::<T, I>::iter() {
+			if !approval_index.contains(&proposal_index) {
+				let err_amount = T::Currency::unreserve(&p.proposer, p.bond);
+				if err_amount.is_zero() {
+					Proposals::<T, I>::remove(proposal_index);
+					log::info!(
+						target: LOG_TARGET,
+						"Released bond amount of {:?} to proposer {:?}",
+						p.bond,
+						p.proposer,
+					);
+				} else {
+					defensive!(
+						"err_amount is non zero for proposal {:?}",
+						(proposal_index, err_amount)
+					);
+					Proposals::<T, I>::mutate_extant(proposal_index, |proposal| {
+						proposal.value = err_amount;
+					});
+					log::info!(
+						target: LOG_TARGET,
+						"Released partial bond amount of {:?} to proposer {:?}",
+						p.bond - err_amount,
+						p.proposer,
+					);
+				}
+				proposals_released += 1;
```

Comment:
may be proposals_precessed? you can also create mutable weight var, and add the weight while you pregress